### DISCUSSION TITLE
Make the guts of LocalProxy.SetApplicationContext re-usable?

### DISCUSSION THREAD

**Discussion by swegele** [MarimerLLC/csla#3591] [Questions] [answered]
I'm pondering about how the actual meat of the method `SetApplicationContext` currently resides in **LocalProxy.cs**.  
I want to reuse that method inside my business objects to swap out AppContextA for AppContextB.  That recursive function already does everything so nicely.  But the method is not in a re-useable place I could access.

Scenario - in memory cache of read-only CSLA objects:
- UserA is the first in the system and triggers a fetch list of MyObjects from the DB.  Results are stored in memory cache
- UserB is second on the system and his same behavior skips the DB fetch to simply return list from cache
- The in memory csla cache service must swap UserA's `ApplicationContext` to be UserB's before giving back the list of MyObjects

Currently I accomplish this by serializing/deserializing the objects to/from byte arrays.  That operation of serialize/deserialize has the side-effect swapping out the `ApplicationContext` inside each object.  But that needlessly incurs the perf hit of going into and out of the byte array.  On large lists this does take a while.  Whereas swapping out `ApplicationContext` is pretty fast.

I'm wondering if we can move the "do it" part of the `SetApplicationContext` method somewhere re-usable in the CSLA framework.  That would empower me to do something like this in my read-only base class:

`public void SwapApplicationContext(ApplicationContext newContext)`

_Obviously it's on me to make sure that the swap in my business objects is allowed and valid from a security perspective_

**Comment by rockfordlhotka** [comment, marked as answer]
Wouldn't this cause a concurrency issue, where multiple users could have access to the object graph at the same time, potentially with the context switching out from underneath them as they interact with the graph?

I think the best way to cache an object graph is to cache the serialized byte array, and deserialize it on request for each user.

**Reply by swegele** [reply]
Ah good point.  The actual ref would no longer be unique coming back from the cache service.  If deserialized then it is new. 

Thanks for helping me think that through.  
bonk - "I should have had a V8"

**Reply by rockfordlhotka** [reply]
You should be able to save 50% of the time by caching the byte array though, so hopefully that helps!

**Reply by swegele** [reply]
yep exactly - `IDistributedCache`
